### PR TITLE
fuzz-testing: euler deposit and withdraw

### PR DESCRIPTION
**Pull request by haythem96**
# Task: euler deposit and withdraw

## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes ENG-840

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Document update

## How Has This Been Tested

Please describe how to test to verify the changes. Provide instructions so we can reproduce.

## FE Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] Added video recordings if it is a UI change

## User Facing Checklist

- [ ] I fully understand the user problem this PR is solving
- [ ] I know who the target user is for this PR and have a deep understanding of that user
- [ ] I have tried this flow thinking from the pov of the target user for this PR
- [ ] I (or working w someone on team) have scheduled a user test for this PR (if it is a large change)

### CODE REVIEW

File: packages/bull-vault/test/fuzz-test/LeverageBullFuzzTest.t.sol [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+        vm.prank(address(leverageBull));
+        IEulerMarkets(eulerMarketsModule).enterMarket(0, weth);
+
+        uint256 balanceOfUnderlying =
+            IEulerEToken(eToken).balanceOfUnderlying(address(leverageBull));
+
+        uint256 ethUsdPrice = UniOracle._getTwap(ethUsdcPool, weth, usdc, TWAP, false);
+        uint256 usdcToBorrow = _wethToDeposit.wmul(ethUsdPrice).wdiv(TARGET_CR).div(WETH_DECIMALS_DIFF);
+        uint256 eulerUsdcBalance = IERC20(usdc).balanceOf(euler);
```
Comment (alpinechicken) [inline packages/bull-vault/test/fuzz-test/LeverageBullFuzzTest.t.sol:174]: why not use dToken.balanceOfUnderling?

**Review by alpinechicken — APPROVED** [review]
lgtm

File: packages/bull-vault/test/fuzz-test/LeverageBullFuzzTest.t.sol [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+        assertEq(usdcBalanceAfter.sub(usdcBalanceBefore), usdcToBorrow);
+
+        vm.startPrank(auction);
+        uint256 usdcToRepay = IEulerDToken(dToken).balanceOf(auction);
+        IERC20(usdc).approve(address(leverageBull), usdcToRepay);
+        leverageBull.auctionRepayAndWithdrawFromLeverage(usdcToRepay, 0);
+
+        assertEq(IEulerDToken(dToken).balanceOf(auction), 0);
+        assertEq(usdcBalanceAfter.sub(IERC20(usdc).balanceOf(auction)), 0);
```
Comment (aleone) [inline packages/bull-vault/test/fuzz-test/LeverageBullFuzzTest.t.sol:194]: i dont get this assertion - aren't we asserting that the usdc balance doesn't change when we repay?

Comment (aleone) [reply packages/bull-vault/test/fuzz-test/LeverageBullFuzzTest.t.sol:194]: how is this passing?

File: packages/bull-vault/test/fuzz-test/LeverageBullFuzzTest.t.sol [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+        vm.startPrank(auction);
+        // deposit first WETH in euler
+        IWETH9(weth).deposit{value: _wethToDeposit}();
+        IWETH9(weth).approve(address(leverageBull), _wethToDeposit);
+        leverageBull.depositAndBorrowFromLeverage(_wethToDeposit, 0);
+        vm.stopPrank();
+
+        vm.prank(address(leverageBull));
+        IEulerMarkets(eulerMarketsModule).enterMarket(0, weth);
```
Comment (aleone) [inline packages/bull-vault/test/fuzz-test/LeverageBullFuzzTest.t.sol:214]: same comment

File: packages/bull-vault/test/fuzz-test/LeverageBullFuzzTest.t.sol [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+
+        assertEq(usdcBalanceAfter.sub(usdcBalanceBefore), usdcToBorrow);
+
+        vm.startPrank(auction);
+        uint256 debtAmount = IEulerDToken(dToken).balanceOf(auction);
+        uint256 usdcToRepay = debtAmount.mul(_percentage).div(100);
+        IERC20(usdc).approve(address(leverageBull), usdcToRepay);
+        leverageBull.auctionRepayAndWithdrawFromLeverage(usdcToRepay, 0);
+
+        assertEq(IEulerDToken(dToken).balanceOf(auction).sub(debtAmount), usdcToRepay);
```
Comment (aleone) [inline packages/bull-vault/test/fuzz-test/LeverageBullFuzzTest.t.sol:241]: isnt this a negative number?  or is something funky going on

debtAmount is the amount of USDC debt we have in total before we do any repayment
dToken.balanceOf(auction) is the debt amount after we repay
isnt the 2nd one by definition less than the first?

File: packages/bull-vault/test/fuzz-test/LeverageBullFuzzTest.t.sol [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+        assertEq(usdcBalanceAfter.sub(IERC20(usdc).balanceOf(auction)), usdcToRepay);
```
Comment (aleone) [inline packages/bull-vault/test/fuzz-test/LeverageBullFuzzTest.t.sol:242]: this seems to be right, but very different from the above assert, not sure why (the one that says change is 0)

File: packages/bull-vault/test/fuzz-test/LeverageBullFuzzTest.t.sol (hunk 4 above)
Comment (aleone) [reply packages/bull-vault/test/fuzz-test/LeverageBullFuzzTest.t.sol:241]: turns out       
```uint256 debtAmount = IEulerDToken(dToken).balanceOf(auction); ```

is returning 0, not sure why yet

File: packages/bull-vault/test/fuzz-test/LeverageBullFuzzTest.t.sol (hunk 5 above)
Comment (aleone) [reply packages/bull-vault/test/fuzz-test/LeverageBullFuzzTest.t.sol:242]: usdcToRepay is getting set to 0 per above debtAmount issue

File: packages/bull-vault/test/fuzz-test/LeverageBullFuzzTest.t.sol (hunk 2 above)
Comment (aleone) [reply packages/bull-vault/test/fuzz-test/LeverageBullFuzzTest.t.sol:194]: turns out usdcToRepay is getting set to 0, so this is passing. In every test we are repaying 0 usdc

**Review by aleone — CHANGES_REQUESTED** [review]
Something funky going on with ```        uint256 debtAmount = IEulerDToken(dToken).balanceOf(auction); ```

File: packages/bull-vault/test/fuzz-test/LeverageBullFuzzTest.t.sol [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+
+        vm.startPrank(auction);
+        // deposit first WETH in euler
+        IWETH9(weth).deposit{value: _wethToDeposit}();
+        IWETH9(weth).approve(address(leverageBull), _wethToDeposit);
+        leverageBull.depositAndBorrowFromLeverage(_wethToDeposit, 0);
+        vm.stopPrank();
+
+        vm.prank(address(leverageBull));
+        IEulerMarkets(eulerMarketsModule).enterMarket(0, weth);
```
[the remark this answers is not in the document]
Comment (haythem96) [reply packages/bull-vault/test/fuzz-test/LeverageBullFuzzTest.t.sol:100]: This PR does not have that code yet

File: packages/bull-vault/test/fuzz-test/LeverageBullFuzzTest.t.sol [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
```
[the remark this answers is not in the document]
Comment (haythem96) [reply packages/bull-vault/test/fuzz-test/LeverageBullFuzzTest.t.sol:167]: same

**Review by aleone — APPROVED** [review]

**Pull request merged**